src/Workspaces/Core/Portable/Workspace/Host/LanguageServerProtocol/ILspWorkspaceRegistrationService.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Host.LanguageServerProtocol
+{
+    /// <summary>
+    /// Allows workspaces to register themselves to be considered when LSP requests come in. Any workspace
+    /// registered will be probed for a matching document/solution which can be given to the request handler
+    /// to operate on.
+    /// </summary>
+    public interface ILspWorkspaceRegistrationService
+    {
+        /// <summary>
+        /// Get all current registrations.
+        /// </summary>
+        ImmutableArray<Workspace> GetAllRegistrations();
+
+        /// <summary>
+        /// Register the specified workspace for consideration for LSP requests.
+        /// </summary>
+        void Register(Workspace workspace);
+
+        /// <summary>
+        /// Unregister the specified workspace.
+        /// </summary>
+        void Unregister(Workspace workspace);
```

Comment:
Do we need to make unregister public?  I can't really think of a scenario where someone would want to specifically unregister outside of dispose.  We could then just handle it automatically in http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Workspace.cs,366

A followup question - maybe this shouldn't be public at all, and Register() should just become a method on Workspace.  We then handle the lifetime properly entirely inside workspace.  I also can't think of a case where someone would need a different implementation of this interface other than the LspWorkspaceRegistrationService

Reply:
I don't think this should be public. Who is the consumer?

---
In reply to: [532856797](https://github.com/dotnet/roslyn/pull/49661#discussion_r532856797) [](ancestors = 532856797)

Reply:
> maybe this shouldn't be public at all, and Register() should just become a method on Workspace

That was one of the options we discussed in that meeting, and I have nothing against it in general.

> Who is the consumer?

This was in response to one of the workspaces (maybe metadata as source?) having a shorter lifespan than others, but it could be moved to an internal call and just handled inside Workspace I guess.